framework/src/main/java/org/checkerframework/common/accumulation/AccumulationTransfer.java

```diff
@@ -0,0 +1,152 @@
+     *
+     * @param node the node whose estimate should be expanded
+     * @param result the transfer result containing the store to be modified
+     * @param values the new accumulation values
+     */
+    public void accumulate(Node node, TransferResult<CFValue, CFStore> result, String... values) {
+        Tree tree = node.getTree();
+        if (tree == null) {
+            return;
+        }
+        AnnotatedTypeMirror oldType = typeFactory.getAnnotatedType(tree);
+        AnnotationMirror newAnno = getCombinedAnno(oldType, Arrays.asList(values));
+        insertIntoStores(result, node, newAnno);
+
+        if (tree.getKind() == Kind.METHOD_INVOCATION) {
+            MethodInvocationNode methodInvocationNode = (MethodInvocationNode) node;
+            Node receiver = methodInvocationNode.getTarget().getReceiver();
+            MethodInvocationTree invokedMethod = (MethodInvocationTree) tree;
+
+            while (receiver != null && typeFactory.returnsThis(invokedMethod)) {
+                // Note that this call doesn't do anything if receiver is a method call
+                // that is not deterministic, though it can still continue to recurse.
+                insertIntoStores(result, receiver, newAnno);
+
+                Tree receiverTree = receiver.getTree();
+                // Possibly recurse: if the receiver is itself a method call,
+                // then we need to also propagate this new information to its receiver
+                // if the method being called has an @This return type.
+                //
+                // Note that we must check for null, because the tree could be
+                // implicit (when calling an instance method on the class itself).
+                // In that case, do not attempt to refine either - the receiver is
+                // not a method invocation, anyway.
+                if (receiverTree == null || receiverTree.getKind() != Tree.Kind.METHOD_INVOCATION) {
+                    // Do not continue, because the receiver isn't a method invocation itself. The
+                    // end of the chain of calls has been reached.
+                    break;
+                }
+
+                invokedMethod = (MethodInvocationTree) receiver.getTree();
+                receiver = ((MethodInvocationNode) receiver).getTarget().getReceiver();
+            }
+        }
+    }
+
+    /**
+     * Unions the values in oldType with the values in newValues to produce a single accumulator
+     * type qualifier.
+     *
+     * @param oldType an annotated type mirror whose values should be included, or null which is
+     *     treated as top
+     * @param newValues new values to include
+     * @return an annotation representing all the values
+     */
+    private AnnotationMirror getCombinedAnno(
+            @Nullable AnnotatedTypeMirror oldType, List<String> newValues) {
+        AnnotationMirror oldAnno;
+        if (oldType == null) {
```

Comment:
Is the logic in these lines similar to that in visitMethodInvocation?  If so, consider abstracting it out.
Or, since getCombinedAnnotation is only called once, you could create a method getAnnotationMirror that calls getAnnotatedType and then calls this logic.  That seems like a logical place for it, and then this method could just take an AnnotationMirror as an argument.

Reply:
I rewrote the logic here to make it clear that they're not quite the same - in the annotated type factory, it's not necessary to check for the annotation mirror being null, but here it is. So I don't think there's any need to abstract the one or two lines they share.